src/Streamly/Streams/StreamD.hs

```diff
@@ -2558,25 +2559,43 @@ data CodePoint = FreshPoint
                | GoFour !Word8
                | Four1 !Word8 !Word8
                | Four !Word8 !Word8 !Word8
-               | YieldPoint !Char
-
-{-# INLINE_NORMAL decodeUtf8 #-}
-decodeUtf8 :: Monad m => Stream m Word8 -> Stream m Char
-decodeUtf8 (Stream step state) = Stream step' (state, FreshPoint)
+               | YieldAndContinue !Char
+               | YieldAndStop !Char
+               | Done
+
+data CodingFailureMode
+    = TransliterateCodingFailure
+    | ErrorOnCodingFailure
+    deriving (Show)
+
+{-# INLINE_NORMAL decodeUtf8With #-}
+decodeUtf8With :: Monad m => CodingFailureMode -> Stream m Word8 -> Stream m Char
+decodeUtf8With cfm (Stream step state) = Stream step' (state, FreshPoint)
   where
+    {-# INLINE transliterateOrError #-}
+    {-# INLINE inputUnderflow #-}
+    transliterateOrError e s =
+        case cfm of
+            ErrorOnCodingFailure -> error e
+            TransliterateCodingFailure -> Skip (s, YieldAndContinue '\xFFFD')
+    inputUnderflow s = case cfm of
+       ErrorOnCodingFailure -> error "Input Underflow"
```

Comment:
Should we use something like "decodeUtf8With: Not enough input to decode."? Ideally, we should use "Streamly.String.decodeUtf8" or "Streamly.String.decodeUtf8Lenient" as identification but for that we will have to pass the identification string here.